### PR TITLE
[New Feature] Add support of TransE for link prediction.

### PR DESCRIPTION
**Pull request by Oxfordblue7**
*Issue #, if available:*

*Description of changes:*
Add TransE score function for link prediction.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: tests/end2end-tests/graphstorm-lp/test.sh [hunk 1]
```diff
@@ -156,9 +156,33 @@ python3 -m graphstorm.run.gs_link_prediction --workspace $GS_HOME/training_scrip
 
 error_and_exit $?
 
+echo "**************dataset: Movielens, RGCN layer 2, node feat: fixed HF BERT & sparse embed, BERT nodes: movie, inference: full-graph, negative_sampler: joint, decoder: RotatE, exclude_training_targets: false"
+python3 -m graphstorm.run.gs_link_prediction --workspace $GS_HOME/training_scripts/gsgnn_lp --num-trainers $NUM_TRAINERS --num-servers 1 --num-samplers 0 --part-config /data/movielen_100k_lp_train_val_1p_4t/movie-lens-100k.json --ip-config ip_list.txt --ssh-port 2222 --cf ml_lp.yaml --fanout '10,15' --num-layers 2 --use-mini-batch-infer false --lp-decoder-type rotate --train-etype user,rating,movie movie,rating-rev,user --num-epochs 1 --eval-frequency 300
```
Comment (classicsong) [inline tests/end2end-tests/graphstorm-lp/test.sh:160]: --lp-decoder-type rotate -> --lp-decoder-type transe?

File: tests/unit-tests/test_config.py [hunk 2]
```diff
@@ -1000,7 +1000,7 @@ def create_lp_config(tmp_path, file_name):
         "exclude_training_targets": "error",
         "reverse_edge_types_map": "query,exactmatch,rev-exactmatch,asin",
         "lp_loss_func": "unknown",
-        "lp_decoder_type": "transe",
+        "lp_decoder_type": "transr",
```
Comment (classicsong) [inline tests/unit-tests/test_config.py:1003]: you also need to check whether transe decoder can be correctly parsed.

**Review by classicsong — COMMENTED** [review]
You also need to update the `docs/source/advanced/link_prediction.rst`, `graphstorm.model.rst` and `configuration-run.rst`.
You can do it in a separate PR.

File: tests/end2end-tests/graphstorm-lp/test.sh (hunk 1 above)
Comment (Oxfordblue7) [reply tests/end2end-tests/graphstorm-lp/test.sh:160]: It's for rotate. I added a new test for default rotate (not contrastive or adversarial).

File: python/graphstorm/eval/utils.py [hunk 3]
```diff
@@ -378,6 +378,151 @@ def calc_rotate_neg_tail_score(heads, tails, r_emb, num_chunks,
...
+        t_emb = t_emb.to(device)
+
+    hidden_dim = h_emb.shape[1]
+    h_emb = h_emb.reshape(num_chunks, neg_sample_size, hidden_dim)
+    t_emb = t_emb - r_emb
+    t_emb = t_emb.reshape(num_chunks, chunk_size, hidden_dim)
+
+    if norm == 'l1':
+        transe_score = gamma - th.cdist(t_emb, h_emb, p=1)
+    else:
```
[the remark this answers is not in the document]
Comment (Oxfordblue7) [reply python/graphstorm/eval/utils.py:476]: The score_norm has been asserted in config/, so norm will only be one of 'l1' or 'l2'.

File: python/graphstorm/config/argument.py [hunk 4]
```diff
@@ -643,6 +644,7 @@ def verify_link_prediction_arguments(self):
         _ = self.lp_loss_func
         _ = self.lp_decoder_type
         _ = self.gamma
+        _ = self.score_norm
```
[the remark this answers is not in the document]
Comment (Oxfordblue7) [reply python/graphstorm/config/argument.py:647]: I was trying to declare a general argument that can be used for other score functions in the future.

File: python/graphstorm/config/argument.py [hunk 5]
```diff
@@ -3149,6 +3149,12 @@ def _add_link_prediction_args(parser):
             default=argparse.SUPPRESS,
             help="Common hyperparameter symbol gamma."
     )
+    group.add_argument(
```
Comment (classicsong) [inline python/graphstorm/config/argument.py:3152]: Do we still need this?

**Review by classicsong — COMMENTED** [review]
One small comment.
Overall LGTM.

**Review by classicsong — APPROVED** [review]

File: python/graphstorm/config/argument.py (hunk 5 above)
Comment (Oxfordblue7) [reply python/graphstorm/config/argument.py]: deleted

**Pull request merged**